### PR TITLE
 Make extended drop table code available for mods to use for craftitems and tools.

### PR DESCRIPTION
**Pull request by ashtrayoz**
This is almost entirely just refactoring existing code.  There are a few minor tidy ups.

### CODE REVIEW

File: doc/lua_api.txt [hunk 1]
```diff
@@ -3402,8 +3402,10 @@ You can find mod channels communication scheme in `docs/mod_channels.png`.
     * Returns `nil` if the given `paramtype2` does not contain color
       information.
 * `minetest.get_node_drops(nodename, toolname)`
-    * Returns list of item names.
+    * Returns list of itemstrings.
     * **Note**: This will be removed or modified in a future version.
+* `minetest.get_extended_drops(droptable, toolname, palette_index)
+    * Returns list of itemstrings.
```
Comment (SmallJoker) [inline doc/lua_api.txt]: This function needs a description of what it does, something like: Calculates the actual items to drop from the given droptable.

File: doc/lua_api.txt [hunk 2]
```diff
@@ -5384,8 +5421,8 @@ Definition tables
         ^ Name of dropped node when dug. Default is the node itself.
         ^ Alternatively:
         drop = {
-            max_items = 1,  -- Maximum number of items to drop.
-            items = {  -- Choose max_items randomly from this list.
+            max_choices = 1,  -- Maximum number of items to drop.
+            choices = {  -- Choose max_choices randomly from this list.
```
Comment (SmallJoker) [inline doc/lua_api.txt]: Why this change? Items as in ItemStacks is a correct term here.

File: doc/lua_api.txt [hunk 3]
```diff
@@ -3402,8 +3402,45 @@ You can find mod channels communication scheme in `docs/mod_channels.png`.
...
       information.
 * `minetest.get_node_drops(nodename, toolname)`
-    * Returns list of item names.
+    * Returns list of itemstrings.
     * **Note**: This will be removed or modified in a future version.
+* `minetest.get_extended_drops(drop_table, toolname, palette_index)
+    * Calculates the item(s) to drop, given the drop_table and the tool.
+    * palette_index is optional, see Hardware coloring for more detail.
+    * drop_table looks like:
+        drop_table = {
```
Comment (SmallJoker) [inline doc/lua_api.txt]: Should be formatted as multiline code. I.e.
\`\`\`
multiline
code
\`\`\`
This should also be moved to the bottom of this file and referenced as "Drop definition table" (or similar) here and in "Node definition". Otherwise we're documenting the same thing two times.

File: doc/lua_api.txt (hunk 2 above)
Comment (ashtrayoz) [reply doc/lua_api.txt]: Some people (including Shara) have commented that having items inside items gives them headaches.  Actually, the comment on 5424 is not quite right - it is the maximum number of choices that will be made.  If sand is chosen, there will be 2 item strings returned.

File: builtin/game/item.lua [hunk 4]
```diff
@@ -173,41 +173,23 @@ function core.strip_param2_color(param2, paramtype2)
...
+	-- Compatibility with old (0.4) behaviour:
+	--	allow drop_table.max_items and drop_table.items.items,
+	--	but prefer drop_table.max_choices and drop_table.choices.items.
+	local choices = drop_table.choices or drop_table.items
+	if type(choices) ~= "table" then
+		return {}
+	end
+	local max_choices = drop_table.max_choices or drop_table.max_items
+	if type(max_choices) ~= "number" then
+		max_choices = 1
```
Comment (SmallJoker) [inline builtin/game/item.lua]: nil value previously meant "no limit". Fall back to `#choices` instead.

File: doc/lua_api.txt (hunk 2 above)
Comment (Ezhh) [reply doc/lua_api.txt]: Doesn't necessarily mean it should be changed. How many mods will this affect?

**Comment by paramat** [comment]
> This is almost entirely just refactoring existing code. There are a few minor tidy ups.

Problem is, the refactoring will cause many conflicts, delaying the merge of the possibly useful features. And refactors are a headache to review.
Perhaps split off the features to merge first and make refactoring a separate later PR?

**Comment by Zughy** [comment]
I suggest to close this PR, the author hasn't been active on GitHub since two days after the PR has been made. The rebase has been lying there for more than two years
